packages/openapi-code-generator/src/typescript/common/type-builder.spec.ts

```diff
@@ -403,21 +406,206 @@ describe.each(
+      const {code} = await getActualFromModel(
+        ir.intersection({
+          schemas: [
+            ir.object({properties: {a: ir.string()}}),
+            ir.intersection({
+              schemas: [
+                ir.object({properties: {b: ir.string()}}),
+                ir.object({properties: {c: ir.string()}}),
+              ],
+            }),
+          ],
+        }),
+      )
+
+      expect(code).toMatchInlineSnapshot(`
+          "declare const x: {
+            a?: string
+          } & {
+            b?: string
+          } & {
+            c?: string
+          }"
+        `)
+    })
+
+    it("can handle intersecting an object with a union A & (B | C)", async () => {
+      const {code} = await getActualFromModel(
+        ir.intersection({
+          schemas: [
+            ir.object({
+              properties: {base: ir.string()},
+              required: ["base"],
+            }),
+            ir.union({
+              schemas: [
+                ir.object({
+                  properties: {a: ir.number()},
+                  required: ["a"],
+                }),
+                ir.object({
+                  properties: {a: ir.string()},
+                  required: ["a"],
+                }),
+              ],
+            }),
+          ],
+        }),
+      )
+
+      expect(code).toMatchInlineSnapshot(`
+          "declare const x: {
+            base: string
+          } & (
+            | {
+                a: number
+              }
+            | {
+                a: string
+              }
+          )"
+        `)
+    })
+
+    it("can handle intersecting an union with a union (A | B) & (D | C)", async () => {
+      const {code} = await getActualFromModel(
+        ir.intersection({
+          schemas: [
+            ir.union({
+              schemas: [
+                ir.object({
+                  properties: {a: ir.number()},
+                  required: ["a"],
+                }),
+                ir.object({
+                  properties: {a: ir.string()},
+                  required: ["a"],
+                }),
+              ],
+            }),
+            ir.union({
+              schemas: [
+                ir.object({
+                  properties: {a: ir.number()},
+                  required: ["b"],
+                }),
+                ir.object({
+                  properties: {a: ir.string()},
+                  required: ["b"],
+                }),
+              ],
+            }),
+          ],
+        }),
+      )
+
+      expect(code).toMatchInlineSnapshot(`
+          "declare const x: (
+            | {
+                a: number
+              }
+            | {
+                a: string
+              }
+          ) &
+            (
+              | {
+                  a?: number
+                }
+              | {
+                  a?: string
+                }
+            )"
+        `)
+    })
+  })
+
+  async function getActualFromModel(
```

Comment:
Plan to later properly flesh out a set of test cases using in-line `IRModel`'s directly, for cases where `$ref` isn't required.